File: Steven Furnari/mta_1712/js/mta.js
console.log(`MTA Trip Planner - mta.plantrip(startLine, startStation, endLine, endStation)`)

const mta = {
    stations: {
        N: [
            'Times Square',
             '34th', 
             '28th', 
             '23rd', 
             'Union Square', 
             '8th'
            ],
        L: [
            '8th', 
            '6th', 
            'Union Square', 
            '3rd', 
            '1st'
        ],
        '6': [
            'Grand Central', 
            '33rd', 
            '28th', 
            '23rd', 
            'Union Square', 
            'Astor Place'
        ]
    }, // mta.stations

    planTrip: function(startLine, startStation, endLine, endStation){
        const firstLine = this.stations[startLine]; // startLine key        
        const secondLine = this.stations[endLine]; // endLine key
        const start = firstLine.indexOf(startStation); // index of startStation
        const end = secondLine.indexOf(endStation); // index of endStation
        const unionSquareStart = firstLine.indexOf(`Union Square`); // index of Union Square on startLine
        const unionSquareEnd = secondLine.indexOf(`Union Square`) // index of Union Square on endLine
        let stops = 0;
        let trip = [];

        if (startLine === endLine){ // if trip is on the same line            
            for (let i = start+1; i < end+1; i++){ // going forward
                stops++;
                trip.push(firstLine[i]);
            };
            for (let i = start-1; i > end-1; i--){ // going back
                stops++;
                trip.push(firstLine[i]);
            };
            console.log( `You will have ${(stops)} stops on your journey.`);
            console.log( `Starting at ${startStation} and stopping at following stations: ${trip.join(`, `)}` );

        } else if (startLine !== endLine) { // trip swaps line at Union Square
            for (let i = start+1; i < unionSquareStart+1; i++){ // counts stops to union square going forward
                stops++;
                trip.push(firstLine[i]);
            }; 
            for (let i = start-1; i > unionSquareStart-1; i--){ // counts stops to union square going back
                stops++;
                trip.push(firstLine[i]);
            }; 
            for (let i = unionSquareEnd+1 ; i < end+1; i++){ // counts stops on new line from union square going forward
                stops++;               
                trip.push(secondLine[i]);
            };
            for (let i = unionSquareEnd-1 ; i > end-1; i--){ // counts stops on new line from union square going back
                stops++;               
                trip.push(secondLine[i]); 
            }; 
            console.log( `You will have ${(stops)} stops on your journey.`);
            console.log( `Starting at ${startStation} on the ${startLine} line, with a change at Union Square to the ${endLine} line, stopping at following stations: ${trip.join(`, `)}` ); 
        };
    }, // planTrip()
}; // mta{}

// mta.planTrip(`6`, `Grand Central`, '6', `Astor Place`) // going forward on single line
// mta.planTrip(`6`, `Astor Place`, '6', `Grand Central`) // going back on single line
// mta.planTrip(`N`, `Times Square`, 'L', `1st`) // going forward with a change
// mta.planTrip(`L`, `1st`, '6', `Grand Central`) // going back with a change
// mta.planTrip(`N`, `Times Square`, '6', `Grand Central`) // going forward and changing to back
// mta.planTrip(`L`, `1st`, '6', `Astor Place`) // going back and changing to forward

File: Steven Furnari/rubymta_1801/mta.rb
def stations
    {
        "N" => [
            'Times Square',
            '34th', 
            '28th', 
            '23rd', 
            'Union Square', 
            '8th'
        ],

        "L" => [
            '8th', 
            '6th', 
            'Union Square', 
            '3rd', 
            '1st'
        ],
        
        "6" => [
            'Grand Central', 
            '33rd', 
            '28th', 
            '23rd', 
            'Union Square', 
            'Astor Place'
        ]
    }
end

def n
    stations["N"]
end

def l
    stations["L"]
end

def six
    stations["6"]
end

def plan_trip start_line, start_station, end_line, end_station
    start_index = start_line.index start_station # index of start station
    end_index = end_line.index end_station # index of end station
    us_index_start = start_line.index 'Union Square' # index of union square on start line
    us_index_end = end_line.index 'Union Square' # index of union square on end line

    if start_line == end_line # if single line
        def message array, station
            puts "You will have #{array.length} stops on your journey."
            puts "Starting at #{station}, take the following stops: #{array.join(', ')}"
        end 
        
        if start_index < end_index # if single line going forward
            trip_arr = start_line[(start_index+1)..end_index]
            message trip_arr, start_station

        else # if single line going back
            trip_arr = start_line[end_index..(start_index-1)].reverse
            message trip_arr, start_station
        end 

    else # line will swap at union square
        def swap_message array, station
            puts "You will have #{array.length} stops on your journey."
            puts "Starting at #{station} with a change at Union Square, take the following stops: #{array.join(', ')}"
        end

        if start_index < us_index_start && end_index > us_index_end # going forward on both lines   

            trip_arr = start_line[(start_index+1)..us_index_start] + end_line[(us_index_end+1)..end_index]
            swap_message trip_arr, start_station
        
        elsif start_index > us_index_start && end_index < us_index_end # going back on both lines

            trip_arr = start_line[us_index_start..start_index-1].reverse + end_line[end_index..us_index_end-1].reverse
            swap_message trip_arr, start_station
        
        elsif start_index < us_index_start && end_index < us_index_end # going forward to union square, then back

            trip_arr = start_line[(start_index+1)..us_index_start] + end_line[end_index..us_index_end-1].reverse
            swap_message trip_arr, start_station
        
        else                                                           # going back to union square, then forward
            trip_arr = start_line[us_index_start..start_index-1].reverse + end_line[(us_index_end+1)..end_index]
            swap_message trip_arr, start_station

        end
    end 
end


plan_trip six, 'Grand Central', six, 'Astor Place' # going forward on single line

plan_trip six, 'Astor Place', six, 'Grand Central' # going back on single line

plan_trip n, 'Times Square', l, '1st' # going forward with a change

plan_trip l, '1st', six, 'Grand Central' # going back with a change

plan_trip n, 'Times Square', six, 'Grand Central' # going forward and changing to back

plan_trip l, '1st', six, 'Astor Place' # going back and changing to forward